projects/RabbitMQ.Client/client/impl/ChannelBase.cs

```diff
@@ -958,7 +1003,50 @@ public ValueTask BasicPublishAsync<TProperties>(CachedString exchange, CachedStr
             }
 
             var cmd = new BasicPublishMemory(exchange.Bytes, routingKey.Bytes, mandatory, default);
-            return ModelSendAsync(in cmd, in basicProperties, body);
+            using Activity sendActivity = RabbitMQActivitySource.Send(routingKey.Value, exchange.Value, body.Length);
+            BasicProperties props = PopulateActivityAndPropagateTraceId(basicProperties, sendActivity);
+
+            if (sendActivity != null)
+            {
+                await ModelSendAsync(in cmd, in props, body);
+                return;
+            }
+
+            await ModelSendAsync(in cmd, in basicProperties, body);
+        }
+
+        private static BasicProperties PopulateActivityAndPropagateTraceId<TProperties>(TProperties basicProperties, Activity sendActivity) where TProperties : IReadOnlyBasicProperties, IAmqpHeader
+        {
+            BasicProperties props = default;
+            // This activity is marked as recorded, so let's propagate the trace and span ids.
+            if (sendActivity?.ActivityTraceFlags == ActivityTraceFlags.Recorded)
+            {
+                if (sendActivity.IsAllDataRequested)
+                {
+                    if (!string.IsNullOrEmpty(basicProperties.CorrelationId))
+                    {
+                        sendActivity.SetTag(RabbitMQActivitySource.MessageConversationId, basicProperties.CorrelationId);
+                    }
+                }
+
+                props = default;
+                if (basicProperties is BasicProperties properties)
+                {
+                    props = properties;
+                }
+                else if (basicProperties is EmptyBasicProperty)
+                {
+                    props = new BasicProperties();
+                }
+
+                var headers = props.Headers ?? new Dictionary<string, object>();
+
+                // Inject the ActivityContext into the message headers to propagate trace context to the receiving service.
+                DistributedContextPropagator.Current.Inject(sendActivity, headers, InjectTraceContextIntoBasicProperties);
```

Comment:
This is something that should be done always (as opposed to only if `ActivityTraceFlags.Recorded`?

That way the trace flags get send to the consumer so they can inherit the parent span's sampling decision. It is typical for the first span in a trace to make a sampling decision used by all child spans when doing head based sampling.

Additionally, downstream systems may still use trace/span id in other non-tracing signals such as logs or exemplars in metrics so flowing that information even if the span isn't being reported to a tracing backend is still expected.

If this isn't done, then the consumer of the message will be starting a new Trace when they consume the message and will make a new sampling decision which may be unexpected and result in partial traces (only spans from that point down).

Reply:
Good point. If we do have an Activity at all we should probably forward that context regardless of the sampling decision. Does that make sense @lmolkova?

Reply:
Set this to always propagate if the `activity != null`